src/cart/crt_swim.c

```diff
@@ -94,27 +103,73 @@ static void crt_swim_srv_cb(crt_rpc_t *rpc_req)
 {
 	struct crt_grp_priv	*grp_priv = crt_gdata.cg_grp->gg_srv_pri_grp;
 	struct crt_swim_membs	*csm = &grp_priv->gp_membs_swim;
+	struct swim_context	*ctx = csm->csm_ctx;
 	struct crt_rpc_swim_in	*rpc_swim_input = crt_req_get(rpc_req);
-	int			 rc;
+	struct crt_rpc_swim2_out *rpc_swim_output = crt_reply_get(rpc_req);
+	swim_id_t		 self_id = swim_self_get(ctx);
+	int			 rc = -ENOTCONN;
 
 	D_ASSERT(crt_is_service());
 
-	D_DEBUG(DB_TRACE, "incoming message %lu <= %lu\n",
-		swim_self_get(csm->csm_ctx), rpc_swim_input->src);
+	D_TRACE_DEBUG(DB_TRACE, rpc_req,
+		"incoming opc %#x with %zu updates %lu <= %lu\n",
+		rpc_req->cr_opc, rpc_swim_input->upds.ca_count,
+		self_id, rpc_swim_input->src);
+
+	if (self_id != SWIM_ID_INVALID) {
+		rc = swim_parse_message(ctx, rpc_swim_input->src,
+					rpc_swim_input->upds.ca_arrays,
+					rpc_swim_input->upds.ca_count);
+		if (rc == -ESHUTDOWN)
+			swim_self_set(ctx, SWIM_ID_INVALID);
+		else if (rc)
+			D_ERROR("swim_parse_message() failed rc=%d\n", rc);
+	}
+
+	if (rpc_req->cr_opc & 0xFFFF) {
+		D_TRACE_DEBUG(DB_TRACE, rpc_req,
+			"reply to opc %#x with %zu updates %lu <= %lu rc=%d\n",
+			rpc_req->cr_opc, rpc_swim_input->upds.ca_count,
+			self_id, rpc_swim_input->src, rc);
 
-	rc = swim_parse_message(csm->csm_ctx, rpc_swim_input->src,
-				rpc_swim_input->upds.ca_arrays,
-				rpc_swim_input->upds.ca_count);
-	if (rc)
-		D_ERROR("swim_parse_message() failed rc=%d\n", rc);
+		rpc_swim_output->rc = rc;
+		rc = crt_reply_send(rpc_req);
+		if (rc)
+			D_ERROR("send reply %d failed: rc=%d\n",
+				rpc_swim_output->rc, rc);
+	}
 }
 
+static int crt_swim_get_member_state(struct swim_context *ctx, swim_id_t id,
+				     struct swim_member_state *state);
+static int crt_swim_set_member_state(struct swim_context *ctx, swim_id_t id,
+				     struct swim_member_state *state);
+
 static void crt_swim_cli_cb(const struct crt_cb_info *cb_info)
 {
-	struct crt_rpc_swim_in *rpc_swim_input = crt_req_get(cb_info->cci_rpc);
+	struct swim_context	*ctx = cb_info->cci_arg;
+	crt_rpc_t		*rpc_req = cb_info->cci_rpc;
+	struct crt_rpc_swim_in	*rpc_swim_input = crt_req_get(rpc_req);
+	struct swim_member_state id_state;
+	swim_id_t		 id = rpc_req->cr_ep.ep_rank;
+	int			 rc;
 
-	D_DEBUG(DB_TRACE, "opc: %#x cci_rc: %d\n",
-		cb_info->cci_rpc->cr_opc, cb_info->cci_rc);
+	D_TRACE_DEBUG(DB_TRACE, rpc_req,
+		"complete opc %#x with %zu updates %lu => %lu rc=%d\n",
+		rpc_req->cr_opc, rpc_swim_input->upds.ca_count,
+		rpc_swim_input->src, id, cb_info->cci_rc);
+
+	if (cb_info->cci_rc && (rpc_req->cr_opc & 0xFFFF)) {
```

Comment:
What does rpc_req->cr_opc & 0xFFFF == true mean?

Reply:
This is how I check the ping is with reply or not.